[Feature] 카드 CRUD 초기 기능 구현

sxi8022 opened the pull request:
## 관련 Issue

#12

## 변경 사항

[feature] 카드 CRUD 기능 구현 
1. 카드 수정 : PUT- /api/cards/{cardId}
2. 카드 작업자 변경 : PATCH - /api/cards/{cardId}/members
3. 카드 이동 : PATCH- /api/cards/{cardId}/move
4. 카드 삭제 : Delete- /api/cards/{cardId}
5. 날짜 지정 : Patch - /api/cards/{cardId}/date
6. 사용자별 알람 엔티티 생성
7. aws config 추가 
8. application.properties 수정
9. querydsl config 추가 
추가기능이나 에러 발생 시 수정 예정


## Todo List

- 포스트맨으로 테스트 완료 
## 진행 상황 
![230808_itwassummer_테스트_aws_s3](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/7ce89516-ee5c-4e51-be28-0172bf51c0e9)
![230808_itwassummer_테스트_정렬순서변경](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/84473c84-8266-4750-a5f9-cb6d27b982e6)
![230808_itwassummer_테스트_카드등록첨부파일](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/2e1a8cb8-b897-4df7-939e-239e3314a9a3)

![230808_itwassummer_테스트_카드수정첨부파일](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/4415f2ea-f40f-4276-a2e1-19dfdd1c232d)
![230809_itwassummer_테스트_postman카드별사용자변경](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/65297f52-4789-4e88-97e9-25b3d38cba60)
![image](https://github.com/proLmpa/NBC_2tWasSummer/assets/42510512/5352a35c-988c-460c-89c1-df4c65e10476)

## Check List

<!-- 기능 구현을 다 했다면? --> 

- [v] 포스트맨으로 체크완료
- [v] 등록, 삭제  기능을 최초 개발 후  테스트코드 작성하여 테스트 하였음 
- [v] 정렬순서 변경 로직 구현(추후 개선필요)
- [v] 마감일, 작업자 변경 로직 구현
- [v] 사용자별 알람 엔티티 생성
- [ X]  첨부파일 관련한 기능은 테스트 코드로 작성에 시간이 걸리므로 포스트맨으로 대체  

## 개선 필요 사항 


-[] 파일첨부기능 자체를 테스트 코드로 작성하려면 mockMVC의 기능을 활용해야 한다. 오랜시간이 걸려서 후순위로 미루고 작업 해야할 것 같다.


-[] queryDsl을 작성하여 카드의 정렬 순서를 바꾸는 작업을 구현했습니다.  데이터 건수가 많으면 쿼리를 개선할 필요가 있다. 

- [] 함수 맨뒤에 exception 처리 했던 소스코드 추후 개선 예정

## 트러블 슈팅

- 포스트맨으로 첨부파일 테스트를 할때 requestDto 클래스와 multipart클래스를 동시에 사용하여 전송하면 에러가 발생하는 경우가 있었다. 
-> requestDto 클래스의 어느테이션을 @RequestPart로 변경 

- 기존에 강의를 보면서 구현했었던 파일첨부기능은 엔티티당 하나의 URL을 가지는 파일첨부였다. 
 -> config 파일과 fileUploader라는 이름의 클래스로 파일첨부 기능을 재구성 
 -> 그래서 표시할때는 RESPONSEDTO로 여러가지 파일의 정보를 list형태로 보여주는 방식으로 개선 
 -> 파일이 여러개 첨부가 될때에는 '|' 라는 구분자를 써서 여러개의 url이 db에 문자열로 들어가도록 하였다.

## 참고 
postman 테스트 경로 : https://documenter.getpostman.com/view/9928633/2s9Xy2MrGz


## 공지
슬랙DM으로 AWS 계정정보 공유할예정

sxi8022 committed:
- 카드 엔티티 구현 (다른 테이블과 관계는 추후에 구현)
- 컨트롤러 테스트 코드를 통해 카드 데이터 등록 가능
- 수정 삭제 메소드 작업 중

sxi8022 committed:
Merge branch 'dev' into feature/12/card_crud

# Conflicts:
#	src/main/java/com/example/itwassummer/card/entity/Card.java

sxi8022 committed:
- 날짜 값을 받아올 수 있도록 오류 수정

sxi8022 committed:
[init] QueryDsl 환경설정 추가, jpaRepository 테스트 기법 적용해보기

sxi8022 committed:
- 수정삭제 기능 추가

sxi8022 committed:
[feature] 삭제 기능 테스트 코드 수정

sxi8022 committed:
[feature] 카드 CRUD 기능 구현
카드 수정 : PUT- /api/cards/{cardId}
카드 작업자 변경 : PATCH - /api/cards/{cardId}/members
카드 이동 : PATCH- /api/cards/{cardId}/move
카드 삭제 : Delete- /api/cards/{cardId}
날짜 지정 : Patch - /api/cards/{cardId}/date
추가기능이나 에러 발생 시 수정 예정

proLmpa commented on src/main/java/com/example/itwassummer/card/entity/Card.java:
```diff
@@ -14,4 +18,43 @@ public class Card {
+
+    @Column(length = 10, nullable = false)
+    private Long parentId;
+
```
length 가 사이즈 제한인가요?

sxi8022 committed:
[bug] s3 프로포티 공유하지 않도록 설정

proLmpa commented on src/main/java/com/example/itwassummer/card/service/CardServiceImpl.java:
```diff
@@ -4,46 +4,225 @@
+
+  private final CardMembersRepository cardMembersRepository;
+
+  /**
```
이미 interface에 적어놓은 거라 Impl에서 까지 다시 적어놓을 필요는 없을 것 같아요
아래 메소드들 위에 적은 것들도 같이 지워도 됩니다.

proLmpa replied:
ServiceImpl의 각 메소드 위에 달아놓은 설명 주석을 얘기하는 겁니다!

sxi8022 replied:
네

sxi8022 replied:
네 impl 쪽의 주석을 지우기로 하였습니다.

proLmpa commented on src/main/java/com/example/itwassummer/card/service/CardServiceImpl.java:
```diff
@@ -4,46 +4,225 @@
+    // 파일첨부
+    List<S3FileDto> fileDtoList = null;
+    String attachmentList = "";
+    if (files != null) {
```
requestDto에 파일이 있는 경우에 정상적으로 변경되는 로직인데, requestDto에 파일이 없으면 기존에 있던 파일은 그대로 저장되어 있나요?

sxi8022 replied:
기존에 그대로 저장되어있는 구조였는데, 
파일 수정시에는 전부 삭제하고 다시 등록하는 구조로 변경하였습니다. 
김진성 튜터님은 파일 등록할때 url 만 저장하지말고 
파일경로하고, 파일명도 json으로 묶어서 db에 컬럼값으로 저장하라고 하시더라고요.(데이터 타입은 text)
별도의 이슈  생성해서 개선방안으로 알아봐야할것 같습니다.

proLmpa commented on src/main/java/com/example/itwassummer/card/dto/CardRequestDto.java:
```diff
@@ -30,7 +33,10 @@ public class CardRequestDto {
 
-
-
+  // 덱 id (외래키)
```
dto에서 entity로 변환할 때 이건 사용하지 않아서 지워도 될 것 같습니다.
그리고 외래키 설정은 dto에서 지정하기 보다 service에서 id를 통해 찾은 deck을 card 엔티티에 매핑된 deck에 저장하는 방식이라 Deck CRUD가 완성되면 그때 만드시면 될 것 같아요

proLmpa replied:
requestDto에 있는 deckId를 말씀드린 겁니다.
진짜 제대로 피드백 하는 건 처음이라 자꾸 이상하거나 쓸데없는 부분을 잘라왔네요

proLmpa commented on src/main/java/com/example/itwassummer/card/service/CardServiceImpl.java:
```diff
@@ -38,12 +39,32 @@ public CardResponseDto save(CardRequestDto requestDto) throws IllegalAccessExcep
+  // 공통 에러메시지는 추후에 커밋된 내용 받아서 수정 예정
+  public Card findCard(Long cardId) {
+    return cardRepository.findById(cardId).orElseThrow(() ->
+        new RuntimeException("게시글이없습니다.", null));
```
예외가 발생하면 에러 메시지가 포스트맨에 출력되나요?
안 된다면 CustomErrorCode에 새로운 에러 메시지 하나 만드시고 CustomException으로 던지면 바로 출려되게 설정해놨어요 :D

sxi8022 replied:
해당 부분 커스텀에러코드에 추가하여 커밋했습니다.

proLmpa commented on src/main/java/com/example/itwassummer/cardmembers/dto/CardMembersResponseDto.java:
```diff
@@ -0,0 +1,23 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class CardMembersResponseDto {
+  // 사용자 이메일
```
이거 camel 스타일로 바꿔주실 수 있나요? (UserEmail -> email, NickName -> nickname, CardName -> cardName)

JisooPyo replied:
이럴 때는 15~21 쭉 드래그 해서 +로 코드 리뷰 해주시면 conversation에서 한 눈에 볼 수 있답니다 ~😁😁

JisooPyo commented on src/main/java/com/example/itwassummer/card/entity/Card.java:
탭이 두 칸이시군요..! 탭 차이가 이렇게 큰 change를 만들었군요..! 😮😮😮 ㅋㅋㅋㅋ

sxi8022 replied:
자동정렬하다보니 이렇게 되었던 것 같네요

JisooPyo commented on src/main/java/com/example/itwassummer/card/entity/Card.java:
```diff
@@ -18,31 +19,35 @@
-    @Column(length = 10, nullable = false)
-    private Long parentId;
+  @Column(length = 10, nullable = false)
+  private Long parentId;
```
parentId length = 10 수정 안 되어 있습니다..!

proLmpa commented on src/main/java/com/example/itwassummer/cardmembers/entity/CardMembers.java:
```diff
@@ -0,0 +1,32 @@
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  private Long id;
+
+  @ManyToOne(fetch = FetchType.LAZY)
```
각 CardMember 마다 한 명의 user와 한 개의 card를 매핑해주는 것이기 때문에 클래스명과 필드명을 모두 단수형으로 바꿨으면 좋겠습니다.
(CardMembers -> CardMember, users -> user, cards -> card)

proLmpa replied:
이러면 User와 Card에 @OneToMany로 매핑된 필드들의 mappedBy도 단수형으로 바뀌어야 할 겁니다.

proLmpa replied:
아 저희 애초에 class 명은 CardMembers, BoardMembers 처럼 전부 복수형 이었네요...  그냥 필드명만 바꿔주시면 될 것 같아요!

JisooPyo commented on src/main/java/com/example/itwassummer/card/service/CardServiceImpl.java:
CardService에 있는 주석들은 빼도 괜찮지 않을까요..!

sxi8022 replied:
네 그렇게 진행해도 좋을것같습니다.

Wooin-dev replied:
저는 서비스에 주석 넣고 있긴해요. 유지해서 마이너스인 요소가 있을까요?

Wooin-dev replied:
아 CardServiceImpl에는 지워도 될 것 같아요

sxi8022 replied:
참고하겠습니다 impl에 있는 주석들을 지워보도록 하겠습니다.

proLmpa commented on src/main/java/com/example/itwassummer/cardnotifications/entity/CardNotifications.java:
```diff
@@ -0,0 +1,40 @@
+
+  @ManyToOne(fetch = FetchType.LAZY, cascade = {CascadeType.PERSIST, CascadeType.REMOVE})
+  private Card cards;
+
```
여기도 CardMembers에서 드린 피드백과 동일하게 고쳐주세요.

JisooPyo commented on src/main/resources/application-s3.properties:
여기있는 파일은 삭제를 해 놓는게 좋을 것 같아요..! 아마 원격에서 삭제된 채로 머지 되고 그걸 태훈님이 pull해오시면 s3가 없어져있을 수도 있긴 한데 다시 만들면 되어요..!

sxi8022 replied:
알겠습니다

JisooPyo reviewed (commented):
1차 리뷰 끝!

proLmpa reviewed (commented):
첫 피드백입니다

JisooPyo commented:
아 이건 정말 별 거 아니지만 PR 내용에 `- [ ] 체크1` 이런 식으로 쓰셔야 체크박스가 된답니다!
체크박스에 체크하시려면 `- [X] 체크1` 이런 식으로 쓰시면 체크 박스에 체크가 됩니다!

JisooPyo commented:
아! 이것도 정말 별거 아니지만 PR 제목 혹시 바꿔주실 수 있으실까요..!! [feature] <구현한 기능 설명>
이런 식으로..!

sxi8022 committed:
[bug]  s3 환경설정 파일 제거

sxi8022 committed:
[bug] 카드 crud 피드백받고 소스코드 수정
- 주석은 service 인터페이스에 작성 , impl 클래스에서는 제거
- parentId 컬럼의 @Column 속성 변경
- requestDto에 있는 deckId 제거(외래 키 관계는 추후 구현)
- 소스코드 정리
- 카드별 사용자, 카드별 알림 엔티티 단수형으로 변경 (erd반영필요)
- 카드 수정 시 첨부파일 없으면 s3에서 파일 삭제되도록 기능 개선
  추후 소스코드 refactor 필요함

sxi8022 committed:
Merge branch 'dev' into feature/12/card_crud

sxi8022 committed:
[bug] 카드 crud 피드백받고 소스코드 수정
- CustomErrorCode로 통일해서 코드 작성 예정

sxi8022 committed:
[bug] 카드 정렬 순서 이동시 returnDto의 paraentId값은 바꾼값으로 나오도록 처리함

proLmpa reviewed (approved):
거의 모든 class가 복수형으로 작성되어 있어서 단수형으로 바꾸는 게 대공사였을 텐데 다 수정해주셨군요!
감사합니다 :D
